### ISSUE TITLE
Selective Sync

### ISSUE THREAD

**Issue by janl** [hoodiehq/hoodie#152]
Hoodie should provide a nice API and implementation for selectively syncing data from the backend to the frontend for cases where the per-user data exceeds local storage limits.

A concrete use-case would be an email application that has many times more emails than are relevant for any particular view of the actual app. Hoodie should make it easy to sync the relevant data items automatically. Similarly, it should discard items that are no longer needed by user-required functions.

This also means that some functions in the app aren’t available if the online source of selective data isn’t available. The implementation should account for that and provide a nice API to handle that case, too.

**Comment by gr2m** [comment]
**tl;dr**
1. I'd do a selective bootstrap of data, instead of a selective sync
2. Can we use CouchDB's URL-rewrite to do `GET /bootstrap?since=123`, that may or may not be filtered?

---

`hoodie.js` is prepared for a selective bootstrap, not sync. But I think that is sufficient. Right now, the initial bootstrap is loading all current data using `GET /_changes?include_docs=true`, with an additional `since=123` if a data has been pulled before and therefore a seq number is locally cached.

We could change that to `GET /bootstrap?since=123` and use the build-in URL rewrites to point it to `/_changes?include_docs=true&seq=123` by default. And if I need a filtered bootstrap instead, it would point to a custom view, that for example would load only objects that have been changed in the past 2 weeks or so?

Would that be possible?

**Comment by janl** [comment]
Although this is far off, my eventual goal is to be able to do something like gmail in Hoodie, where the selection can change during app use, so I'm not sure if the selective bootstrap works for that case.

**Comment by gr2m** [comment]
for that particular usecase, the selective bootstrap would only load your inbox, not your archived mails. You can work offline by replying, sending or archiving emails. Once back online, it will sync my changes. Sounds pretty straight forward to me. Or am I missing a use case here?

**Comment by janl** [comment]
what happens when the amount of email exceeds local storage? even just the unread bits?

**Comment by janl** [comment]
In addition, I use extensive folders/tags, so I have 100s of “inboxes” all with their current & archived email.

**Comment by gr2m** [comment]
As long as it's just text, we can go pretty far with the offline data, especially after we replaced localStorage with something proper.

There must be something that distinguishes what data you want to be cached offline and what not. We use that for the bootstrap. If you run out of local storage, we get rid of offline data that does not fit the criteria. Once there is no more, we give an error message to inform the user that he needs to remove some stuff from offline storage.

I guess I'm not 100% sure what you mean by "selective sync" or how it differs from the initial bootstrap being filtered? I find it hard to wrap my head around it at the moment, it will be easier once someone actually runs into the problem and we have to find a way to work around it. I know that I will need that pretty soon for minutes, so that might happen rather earlier than later :)

**Comment by janl** [comment]
I agree the topic is still  bit nebulous, but people already ask about what to do if one overflows whatever is useful to store in the browser. Selective bootstrap might be a good enough answer for a good while until we have a use-case where we need something else.

**Comment by gr2m** [comment]
Get input from https://github.com/hoodiehq/hoodie.js/issues/344

**Comment by allain** [comment]
The bootstrap approach using custom views might work, but it'll require users with popular apps to dive deeply into the backend to define them in order to give active users a nice first time load on new boxes. Unless of course, you mean that hoodie defines them and the user never has to see them.

I dislike the idea of having end users know that much about the backend.

**Comment by gr2m** [comment]
We'd provide a nice UI in the admin dashboard for that. A view is just a JavaScript function after all. No CouchDB knowledge will be required
